RecoLocalTracker/SiStripClusterizer/plugins/alpaka/SiStripRawToClusterAlgo.dev.cc

```diff
@@ -0,0 +1,1096 @@
+
+    auto invalidFedChN_d = make_device_buffer<uint32_t>(queue);
+    alpaka::memset(queue, invalidFedChN_d, 0);
+
+    alpaka::exec<Acc1D>(queue,
+                        workDiv,
+                        SiStripKer_applyQualConds{},
+                        nstrips_d.data(),
+                        invalidFedChN_d.data(),
+                        fedBuffer_d_->data(),
+                        stripMapping_d_->view(),
+                        conditions_DetToFeds);
+
+    const uint32_t threads = 1024u;
+    const uint32_t nBlocks = divide_up_by(stripMapping_d_->view().metadata().size(), threads);
+    const auto workDivMultiBlock = make_workdiv<Acc1D>(nBlocks, threads);
+    auto blockCounter_d = make_device_buffer<int32_t>(queue);
+    alpaka::memset(queue, blockCounter_d, 0);
+    alpaka::exec<Acc1D>(queue,
+                        workDivMultiBlock,
+                        multiBlockPrefixScan<uint32_t>(),
+                        stripMapping_d_->const_view().fedChStripsN().data(),
+                        stripMapping_d_->view().fedChStripsN().data(),
+                        (uint32_t)stripMapping_d_->view().metadata().size(),
+                        (int32_t)nBlocks,
+                        blockCounter_d.data(),
+                        alpaka::getPreferredWarpSize(alpaka::getDev(queue)));
+
+    // To do: merge the two kernels for applyQualConds and sum into a single one.
+
+    // Get the number of strips to unpack
+    nStrips_h_ = cms::alpakatools::make_host_buffer<uint32_t>(queue);
+    const uint32_t size = stripMapping_d_->view().metadata().size();
+    auto viewSrc = make_device_view(queue, stripMapping_d_->view().fedChStripsN(size - 1));
+    auto viewDst = make_host_view(*nStrips_h_->data());
+    alpaka::memcpy(queue, viewDst, viewSrc);
+  }
+
+  void SiStripRawToClusterAlgo::unpackStrips(Queue& queue, const Data* conditions_Data) {
+    // std::cout << "#kerStrips," << nStrips_h->data()[0] << std::endl;
+    // std::cout << "#invalidFed," << invalidFedChN_d.data()[0] << std::endl;
+
+    // Allocate the SiStripDigi collection on device
+    const uint32_t nStrips = nStrips_h_->data()[0];
+    digis_d_ = std::make_unique<SiStripDigiDevice>(nStrips, queue);
+
+    // Run the unpacking kernel
+    uint32_t divider = 32u;
+    uint32_t nChannels = (*stripMapping_d_)->metadata().size();
+    uint32_t groups = divide_up_by(nChannels, divider);
+    auto workDiv = make_workdiv<Acc1D>(groups, divider);
+    alpaka::exec<Acc1D>(queue,
+                        workDiv,
+                        SiStripKer_unpackZS2{},
+                        fedBuffer_d_->data(),
+                        digis_d_->view(),
+                        stripMapping_d_->const_view(),
+                        conditions_Data);
+
+    // dumpUnpackedStrips(queue, digis_d_.get()); // (for debugging)
+
+    // Allocate and initialize the StripClustersAux collection
+    sClustersAux_d_ = std::make_unique<StripClustersAuxDevice>(nStrips, queue);
```

Comment:
Instead of allocating and filling this for every event, it could be done once and propagated automatically to all devices using the `MoveToDeviceCache` functionality.

See [the REDAME](https://github.com/cms-sw/cmssw/blob/master/HeterogeneousCore/AlpakaCore/README.md#copy-eg-configuration-data-to-all-devices-in-edproducer) and [an example](HeterogeneousCore/AlpakaTest/plugins/alpaka/TestAlpakaGlobalProducerMoveToDeviceCache.cc).